### PR TITLE
Fold full-width SIMD-typed indirections

### PR DESCRIPTION
**Pull request by SingleAccretion**
I. e. `OBJ/IND<simd>(ADDR(LCL_VAR<simd>))`.

Some positive (in nature) [diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=44369&view=ms.vss-build-web.run-extensions-tab), with a TP win as well.

### CODE REVIEW

**Comment by SingleAccretion** [comment]
@dotnet/jit-contrib 

We're getting close to completing the "fold all indirect access in local morph" part of the `ADDR` work.

**Comment by EgorBo** [comment]
> positive diffs

doesn't sound good 😄 luckily, they're negative in fact!

**Comment by EgorBo** [comment]
Do you want to run some outerloop here? Since over the last few weeks it was constantly in a bad shape we try to run them in all potentially impactful PRs explicitly

**Comment by SingleAccretion** [comment]
> Do you want to run some outerloop here? Since over the last few weeks it was constantly in a bad shape we try to run them in all potentially impactful PRs explicitly

Yeah, seems a good idea ([libraries-]stress would be the pipelines I suppose).

> positive diffs

I have a questionable habit of characterizing diffs by their "niceness" 😄.

**Comment by SingleAccretion** [comment]
I looked at the stress failures, all look pre-existing, except one on ARM:
```

Assert failure(PID 61 [0x0000003d], Thread: 119 [0x0077]): Assertion failed 'cls != nullptr && !isInvalidHandle(cls)' in 'Analysis:<MakeAndAssignEnvironments>b__12_0(Scope):this' during 'Morph - Inlining' (IL size 241; hash 0x4aab3218; FullOpts)

    File: /__w/1/s/src/coreclr/jit/_typeinfo.h Line: 235
    Image: /root/helix/work/correlation/dotnet
```
I looked at the dump, and what we have there is a function being inlined, of which the first parameter is a reference type, but when looking at the signature, we get, rather mysteriously, `CORINFO_TYPE_VALUECLASS`. The method table of this type (as returned by `getArgType`) looked uninitialized (assuming I got the right stack location for it, I suppose):
```
0:025> ?? (MethodTable*)0xd0071090
MethodTable * 0xd0071090
   +0x000 m_dwFlags        : 0xf4884050
   +0x004 m_BaseSize       : 0xd0071c90
   +0x008 m_wFlags2        : 0x113c
   +0x00a m_wToken         : 0xd007
   +0x00c m_wNumVirtuals   : 0xdb5c
   +0x00e m_wNumInterfaces : 0xe1d6
   +0x010 debug_m_szClassName : (null) 
   +0x014 m_pParentMethodTable : (null) 
   +0x018 m_pLoaderModule  : 0xffffffff Module
   +0x01c m_pWriteableData : 0x12340000 MethodTableWriteableData
   +0x020 m_pEEClass       : (null) 
   +0x020 m_pCanonMT       : 0
   +0x024 m_pPerInstInfo   : (null) 
   +0x024 m_ElementTypeHnd : 0
   +0x024 m_pMultipurposeSlot1 : 0
   +0x028 m_pInterfaceMap  : 0xdfc265b0 InterfaceInfo_t
   +0x028 m_pMultipurposeSlot2 : 0xdfc265b0
```
Will assume not related as there are no SIMD types on ARM.

Will also hold off on resolving the conflicts until the CI is up.

**Comment by SingleAccretion** [comment]
Looks like there are a couple SPMI failures that will need to be investigated.

**Comment by SingleAccretion** [comment]
> Looks like there are a couple SPMI failures that will need to be investigated.

The issues have been solved and this is once again ready for review.

File: src/coreclr/jit/morph.cpp [hunk 1]
```diff
@@ -3255,6 +3244,12 @@ GenTreeCall* Compiler::fgMorphArgs(GenTreeCall* call)
                         passingSize  = genTypeSize(structBaseType);
                     }
                 }
+#if !defined(TARGET_X86)
+                else
+                {
+                    hasMultiregStructArgs = true;
```
Comment (jakobbotsch) [inline src/coreclr/jit/morph.cpp:3250]: Why is this needed here now?

File: src/coreclr/jit/morph.cpp [hunk 2]
```diff
@@ -3443,46 +3438,39 @@ GenTreeCall* Compiler::fgMorphArgs(GenTreeCall* call)
...
             GenTree* asgPtrSlot  = gtNewAssignNode(destPtrSlot, argx->AsOp()->gtOp1);
             GenTree* asgTypeSlot = gtNewAssignNode(destTypeSlot, argx->AsOp()->gtOp2);
             GenTree* asg         = gtNewOperNode(GT_COMMA, TYP_VOID, asgPtrSlot, asgTypeSlot);
 
             // Change the expression to "(tmp=val)"
             arg.SetEarlyNode(asg);
-
             call->gtArgs.SetTemp(&arg, tmp);
-            lvaSetVarAddrExposed(tmp DEBUGARG(AddressExposedReason::TOO_CONSERVATIVE));
+            hasMultiregStructArgs |= ((arg.AbiInfo.ArgType == TYP_STRUCT) && !arg.AbiInfo.PassedByRef);
```
Comment (jakobbotsch) [inline src/coreclr/jit/morph.cpp:3455]: I assume this was a preexisting issue on platforms with by-value struct args?

**Review by jakobbotsch — APPROVED** [review]
LGTM, nice clean up too.

File: src/coreclr/jit/morph.cpp (hunk 1 above)
Comment (SingleAccretion) [reply src/coreclr/jit/morph.cpp:3250]: It used to be set below:
```c
#if FEATURE_MULTIREG_ARGS
        if (isStructArg)
        {
            if (((arg.AbiInfo.NumRegs + arg.AbiInfo.GetStackSlotsNumber()) > 1) ||
                (isHfaArg && argx->TypeGet() == TYP_STRUCT))
            {
                hasMultiregStructArgs = true;
            }
        }
```
So this is just moving it to here. The benefit it this location is that we don't need to invoke multi-reg morphing if we have already transformed everything into `FIELD_LIST`s.

File: src/coreclr/jit/morph.cpp (hunk 2 above)
Comment (SingleAccretion) [reply src/coreclr/jit/morph.cpp:3455]: No, as with https://github.com/dotnet/runtime/pull/76745/files#r1009281151, this was just moved.
